parity-clib/parity.h

```diff
@@ -14,166 +14,270 @@
-///	- ud			: Specific user defined data that can used in the callback
-///
-///	- On success	: The function returns an object to the current session
-///					  which can be used cancel the subscription
-///	- On error		: The function returns a null pointer
-///
-void* parity_subscribe_ws(const void *const parity, const char* ws_query, size_t len,
-		void (*subscribe)(void* ud, const char* response, size_t len), void* ud);
+/// @param parity	Reference to the running parity client
+/// @param ws_query JSON encoded string representing the websocket event to
+/// subscribe to.
+/// @param len Length of the query.
+/// @param response Callback to invoke when a websocket event occurs.
+/// @param ud Specific user defined data that can used in the callback.
+/// @param destructor Will be called when `ud` is guaranteed to not be needed
+/// again.
+/// @return A handle to the subscription on success, or NULL on error.
+/// The handle can be used to cancel the subscription.
+struct parity_subscription *parity_subscribe_ws(
+    const struct parity_ethereum *const parity, const char *ws_query,
+    uintptr_t len,
+    void (*subscribe)(void *ud, const char *response, uintptr_t len), void *ud,
+    parity_destructor destructor);
 
-/// Unsubscribes from a websocket subscription. Caution this function consumes the session object and must only be
-/// used exactly once per session.
-///
-///	- session		: Pointer to the session to unsubscribe from
+/// Unsubscribes from a websocket subscription. This function destroys the
+/// session object, leaving `session` a dangling pointer.
 ///
-int parity_unsubscribe_ws(const void *const session);
+/// @param session Pointer to the session to unsubscribe from.  If `session` is
+/// NULL, this is a harmless no-op.
+void parity_unsubscribe_ws(const struct parity_subscription *const session);
 
 /// Sets a callback to call when a panic happens in the Rust code.
 ///
-/// The callback takes as parameter the custom param (the one passed to this function), plus the
-/// panic message. You are expected to log the panic message somehow, in order to communicate it to
-/// the user. A panic always indicates a bug in Parity.
+/// The callback takes as parameter the `void *param` passed to this function
+/// and the panic message. You are expected to log the panic message and exit.
+/// If this function returns, Parity Ethereum will exit the process itself.
 ///
-/// Note that this method sets the panic hook for the whole program, and not just for Parity. In
-/// other words, if you use multiple Rust libraries at once (and not just Parity), then a panic
-/// in any Rust code will call this callback as well.
+/// It is not possible to recover from a panic.  Calling `longjmp` or throwing a
+/// C++ exception results in undefined behavior.  It *is* permissible to block
+/// for an arbitrary amount of time in this callback.  Due to Rust’s memory
+/// safety, it is unlikely that memory has been corrupted, so you can (and
```

Comment:
Data might be in an incorrect state though, so I'd advise to **not** save data to disk before exiting.

Reply:
That is a good point.